### ISSUE TITLE
Having a separate implementation for Canvas is problematic in the longterm.

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/bamboo#46]
As part of Fourier code review https://github.com/phetsims/fourier-making-waves/issues/165, @jonathanolson pointed out that having a separate implementation for Canvas (CanvasChartNode, CanvasLinePlot) is problematic in the longterm.  For example, as we get more support for WebGL to improve performance, code that uses the Canvas-based classes will be stuck with Canvas performance.  An API like Node that chooses its rendered (or allows you to specify a renderer) would be preferrable. 

I'm not sure how we get there from here.

**Comment by pixelzoom** [comment]
Unassigning because I doubt that we'll see progress on this for awhile.  Let me know when you'd like to discuss.

**Comment by samreid** [comment]
This doesn't seem like something we need to solve right away.  All of the canvas charts already have a non-canvas implementation, which could be rendered in SVG or WebGL if scenery one day supports that.  Yes, the canvas implementation is a holdover, but who knows if/when we will have more WebGL support in scenery?  Also, if we have WebGL support in scenery, perhaps we will build or use an adapter that lets us efficiently reuse a lot of our canvas code, like https://github.com/jagenjo/Canvas2DtoWebGL?  

Yes, it's unfortunate that bamboo's canvas and scenery renderers have different APIs, but it doesn't seem practical to unify those APIs eagerly.  But if others argue this is important and should be done soon, perhaps elevate this to epic-level issue? Thoughts @jonathanolson and @pixelzoom?

**Comment by pixelzoom** [comment]
I see @jonathanolson's point about this. If the renderer hadn't been abstracted out eagerly in scenery, there's little chance that we would have been able to do it later, especially if Node (and its subclasses) had different APIs for different renderers.  So if we think it's valuable to do that for bamboo, it should probably be done eagerly.

**Comment by samreid** [comment]
It seems we should add this to our subgroup meeting alongside https://github.com/phetsims/bamboo/issues/45

**Comment by samreid** [comment]
This sounds like an epic-level issue, but one that I won't have time for in 2022 Q1.  @pixelzoom or @jonathanolson can you please comment whether this should be labeled as epic, and how soon it should be worked on?

**Comment by pixelzoom** [comment]
I agree with the concern that @jonathanolson expressed in https://github.com/phetsims/bamboo/issues/46#issue-1004519042.  I don't know whether it should be labeled as an epic, or what the priority should be.   It's not impeding any of my current work.

**Comment by samreid** [comment]
Today @pixelzoom and @jonathanolson and I discussed this.

We saw that CanvasGridLineSet and GridLineSet already have very similar APIs.  But @jonathanolson wanted to see an API like:

```ts
    const horizontalGridLineSet = new GridLineSet( chartTransform, Orientation.HORIZONTAL, 1, gridLineOptions,{
      useCanvas: true
    } );
```

This is not a priority at the moment.  When will it become a priority? Probably when we have a sim that cannot use what is already there.

**Comment by pixelzoom** [comment]
From https://github.com/phetsims/calculus-grapher/issues/256#issuecomment-1459084768, where we identified a need for a Canvas implementation of ScatterPlot:

> * Noting that creating CanvasScatterPlot makes this problem worse: https://github.com/phetsims/bamboo/issues/46. Having 2 parallel sets of bamboo classes is a real problem.  For example, I could not easily switch from LinePlot to CanvasLinePlot in calculus-grapher because the APIs are diffferent -- CanvasLinePlotOptions does not support `lineWidth`.  It would be superior to have one set of bamboo classes, each with a `renderer` option, and have different internal implementations based on the value of `renderer`.  This would ensure a consistent class and Options API.

**Comment by jonathanolson** [comment]
I'd also like to investigate creating better Scenery primitives for this type of thing. We're seeing a lot of cases of wanting high-performance instancing (like sprites), and high-performance changing paths.

If either of those things would be helpful, I'd like to investigate it (as I see those as recurring pain points that could be solved in better ways at the Scenery level).

**Comment by pixelzoom** [comment]
Yes, that would be nice.  Ther current pattern of having parallel Shape-based and Canvas-based implementations is not good.   This will only get more difficult to clean up when more sims are using bamboo (there are currently 7 sims using bamboo.).  Since it's a project you (@jonathanolson) want to tackle, I'll assign back to you to schedule/prioritize.
